### ISSUE TITLE
Unable to Reproduce Reported Results for Qwen2.5-7B-Instruct

### ISSUE THREAD

**Issue by xuchanganan** [idanshen/Self-Distillation#9]
I am currently attempting to reproduce the evaluation results for the Qwen2.5-7B-Instruct model. However, the performance lift I have achieved is minimal, and I would like to verify if my setup or expectations are correct.
1. Configuration Details
Model: Qwen2.5-7B-Instruct (base)
Training Arguments: Default arguments were used.
Epochs: 1
2. Observed Results
I recorded the following eval/rewards scores:
Before Training (Baseline): 0.55882
After 1 Epoch: 0.57353
The improvement is only +0.01471, which seems surprisingly low.
3. Reward
```python
completion_text = completion[-1]["content"] if isinstance(completion, list) else completion
predict_actions = extract_actions(completion_text)
reward = 1.0 if predict_actions == golden_answer else 0.0
```
Thank you.

**Comment by ds5423** [comment]
I confirm the same issue. I found that Qwen2.5-7B-Instruct was already performing very well without any tuning, and the model after self-distill only has marginal improvement. I'd like to kindly request the authors to release the exact test set and evaluation script for reproducibility. @idanshen

**Comment by idanshen** [comment]
I uploaded the checkpoint I got to hf - https://huggingface.co/improbableaimit/sdft-tooluse-7b/tree/main
can you please check it with your eval script?

**Comment by xuchanganan** [comment]
@idanshen 
Thanks for sharing the checkpoint! I tested it on `data/tooluse_data/eval_data.json` (68 samples) and got a reward of `0.54412`.

My script uses strict string matching, so it incorrectly penalizes valid outputs, like giving a 0.0 reward for predicting `language: "en"` when the golden answer is `"English"`.
=====predict_actions=====
 [{'Action': 'getHolidaysByCountry', 'Action_Input': '{"country": "Japan", "year": 2023, "language": "en"}'}]
=====golden_answer=====
 [{'Action': 'getHolidaysByCountry', 'Action_Input': '{"country": "Japan", "year": 2023, "language": "English"}'}]
=====reward=====
 0.0

I assume your reward function is more flexible. To properly align our evaluation, could you please share your eval script or the reward calculation logic?

Thanks!

**Comment by xuchanganan** [comment]
Quick update on my latest experiment:

I switched to `reverse_kl` (`alpha=1.0`) and trained for 2 epochs.
- **Epoch 2:** `eval/rewards` = `0.60294`

While there's an increase, the progress feels slow（2000steps+）. The shape of the reward curve also didn't seem to change much compared to using `forward_kl`.

**Comment by JFernando4** [comment]
I was able to partially reproduce the results shown in Figure 3 using SDFT with the available code and datasets. Here are my results based on a single run:

<img width="1105" height="553" alt="Image" src="https://github.com/user-attachments/assets/fb503904-990f-4c6f-8f85-6638de5ee94b" />

Note that my results are not normalized. The main takeaway from the figure remains valid. When training with SDFT, performance on the Tool Use dataset remains high as the model is trained on the Science dataset. However, the actual performance is slightly lower than originally reported in the paper. In the paper, SDFT achieves about 70% accuracy on both datasets (according to Figure 4), whereas my results reach only about 65% in Tool Use and 60% in Science. 

This is the setup that I used: 
```
{
    "learning_rate", 1e-5,
    "num_train_epochs": 2,
    "num_prompts_per_batch": 32,
    "ref_model_mixup_alpha": 0.02
}
```
In addition, I changed a few parameters in `main.py`.  I changed `warmup_ratio = 0.1` to `warmup_steps = 10` to match Table 3 in the paper. I set `max_completion_length` and `max_prompt_length` to 2048 to match Table 3. I set `alpha = 1.0` instead of the default `0.0` to use reverse KL divergence. 

Lastly, I had to use `paged_adamw_8bit` optimizer from the `bitsandbytes` package because I only have access to H100 GPUs, which struggle to keep the 7B parameter model in memory.

I found that the learning rate of 5e-5 suggested in the README file was too large, causing the loss to explode. Could you please share the exact hyperparameters used in Figure 3? Table 3 in the paper provides only a range of values, not the specific configuration used in the Figure.

**Comment by idanshen** [comment]
Thank you for your help in making sure the results are reproducible! Moving from my own messy codebase to this clean one created more issues than I anticipated. 
For both figures 3 and 4, I used a learning rate of 5e-5, and it worked well.  When I run the command in the README as-is, the loss doesn't explode. For example, for the Science dataset, I got the following model:
https://huggingface.co/improbableaimit/sdft-science-7b

which scores 69% accuracy. Maybe your loss exploded because of the 8bit optimizer?

**Comment by devonbrackbill** [comment]
Hi, I'm trying to reproduce Figure 4 from the paper (retention of baseline capabilities when fine-tuning on Science Q&A). I ran your code exactly on a single H200 (matching Appendix B), lr=5e-5, 2 epochs, Science Q&A, Qwen2.5-7B-Instruct. The directional result reproduces: SDFT forgets less than SFT and learns the new task slightly better. But the magnitude of the forgetting gap is much smaller than Figure 4 suggests, and I have a few questions.

**Our results** (all evals on same machine, max_model_len=4096, lm_eval 0.4.11 `prompt_level_strict_acc` for IFEval):

| Metric | Baseline | SFT lr=5e-5 | SDFT lr=5e-5 | SFT delta | SDFT delta |
|---|---|---|---|---|---|
| HellaSwag | 80.47% | 79.69% | 79.71% | −0.78 | −0.76 |
| MMLU | 71.80% | 70.72% | 71.19% | −1.08 | −0.61 |
| TruthfulQA (mc2) | 64.80% | 60.76% | 62.62% | −4.04 | −2.18 |
| Winogrande | 70.48% | 73.95% | 72.77% | +3.47 | +2.29 |
| IFEval | 57.12% | 32.90% | 37.15% | −24.22 | −19.97 |
| **Prior-task avg** | **68.93%** | **63.60%** | **64.69%** | **−5.33** | **−4.24** |
| **Science Q&A** | **33.33%** | **68.05%** | **69.63%** | **+34.72** | **+36.30** |

Three questions:

1. **Is the ~6 pt SFT vs SDFT forgetting gap in Figure 4 robust?** We see only a ~1 pt gap (−5.33 vs −4.24 avg), with both methods forgetting substantially. Figure 4 shows SDFT staying within ~1 pt of baseline while SFT drops ~7 pts. Was that level of separation consistent?

2. **Which lr produced the Figure 4 Science Q&A point?** The earlier answer here says lr=5e-5, but Table 3 shows a sweep over {5e-6, 1e-5, 5e-5}. At lr=5e-5 both methods show significant IFEval degradation. Would a lower LR produce the larger SDFT vs SFT separation shown in the figure?

3. **Which lm_eval version and IFEval metric were used?** We used lm_eval 0.4.11, `prompt_level_strict_acc`. Our baseline Qwen2.5-7B-Instruct scores 57% on IFEval (`prompt_level_strict_acc`). The paper appears to show a higher baseline (~65-70%). A different lm_eval version or metric variant (e.g., `inst_level_strict_acc`) would explain much of the gap.

**Comment by devonbrackbill** [comment]
Similar to @yanlai00 in [this comment](https://github.com/idanshen/Self-Distillation/issues/9#issuecomment-4077551438), I'm also not able to show forgetting for SFT. 

In this figure, I used Qwen 2.5-7B-Instruct, lr=1e-5, 2 epochs, and the sequential learning setup where I first trained Tool Use (TU) and then Science Q&A. I didn't run the eval on Tool Use while it was training, so you see a jump from the baseline to training step 250. But the blue SFT line doesn't forget TU during Phase 2 Science Q&A training. 

Is this consistent with your parameter sweeps? Or am I implementing something wrong? 

From [this comment](https://github.com/idanshen/Self-Distillation/issues/9#issuecomment-4075784793), I see that you're using a higher lr=5e-5. Did you see SFT forgetting at a lower LR?

<img width="526" height="402" alt="Image" src="https://github.com/user-attachments/assets/0bb47141-c4b4-463b-a2b5-514683eeb0e3" />

**Comment by unsosp1ro** [comment]
> Thank you for your help in making sure the results are reproducible! Moving from my own messy codebase to this clean one created more issues than I anticipated. For both figures 3 and 4, I used a learning rate of 5e-5, and it worked well. When I run the command in the README as-is, the loss doesn't explode. For example, for the Science dataset, I got the following model: https://huggingface.co/improbableaimit/sdft-science-7b
> 
> which scores 69% accuracy. Maybe your loss exploded because of the 8bit optimizer?

---

Hi there, thanks for releasing the model ckpts!
I noticed that the tool-use performance of  [sdft-science-7b](https://huggingface.co/improbableaimit/sdft-science-7b) is not very high, which seems somewhat inconsistent with the paper’s claim about avoiding forgetting. I'm wondering whether this model was obtained by continued training from [sdft-tooluse-7b](https://huggingface.co/improbableaimit/sdft-tooluse-7b), or whether it was trained directly on the science data starting from Qwen2.5-7B-Instruct. Here are full scores (all models were evaluated with `max_new_tokens=2048` and `temperature=0`):

| Model | Tool Use | Sci |
|---|---:|---:|
| Qwen2.5-7B-Instruct (baseline) | 41.2 | 31.2 |
| sdft-tooluse-7b | 70.1 | 34.3 |
| sdft-science-7b | 48.5 | 67.5 |

Could you please clarify the training path for [sdft-science-7b](https://huggingface.co/improbableaimit/sdft-science-7b)? Thanks!
